### PR TITLE
Support goal alerts for all line/area/bar questions

### PR DESCRIPTION
**Pull request by attekei**
**I merged https://github.com/metabase/metabase/pull/6749 to this PR and transferred its todo items here.**

**Frontend TODO:**

- [x] Fallback to "rows present" alert instead of goal alert if a line/area/bar visualization has multiple y-axis columns. Also show a heads-up message in the alert creation dialog in this scenario, like this:
screen shot 2018-01-30 at 21 11 21

- [x] From #6749: We need a migration path for older questions, i.e. if you create a goal alert for an existing question, we require to save a version of a question where the x and y columns are present in visualization_settings. 

- [ ] From #6749: Testing. One thing is that there might (?) be scenarios where changing the query / visualization type in the query builder doesn't properly reset the visualization settings and then old x axis and y axis columns are still in the card object although they shouldn't be there anymore.

---

A potential fix to #6508. In alerts spec we had defined that goal alerts are only available for "timeseries questions" and that only applies to questions with a date breakout. This loosens the criterion so that goal alerts are available regardless of the question type as far as it's using line/area/bar visualization with a goal.

Backend might not support all non-timeseries question types (?) so I'm not sure how well this approach works.

### CODE REVIEW

**Comment by attekei** [comment]
We talked for a while @senior about this PR.

We concluded that even its current state this PR is definitely a step forward from what we've had earlier. The restriction that alerts should be only available for questions with time breakout is completely unnecessary. Simply removing that restriction (what I did in my first commit) improves the user experience remarkably.

There are currently some scenarios in which the alerts don't work correctly / the user experience isn't good. Those are listed below with our current opinions whether they are important or not and how those could be fixed. We'd like to hear your perspectives @maz and @sameer on these!

---

**When using an unaggregated column in x-axis, the frontend sums the values of distinct column values but backend doesn't know about that**
 
It's indicated by the warning sign in visualization settings:

screen shot 2018-01-17 at 16 12 39

@senior thought that while its possible to duplicate this summation logic to the backend, the gains maybe don't outweigh the required effort and risks related to the implementation.

**When having multiseries with two Y axes it might be ambiguous which column the goal line is attached to**
screen shot 2018-01-17 at 16 59 04

We weren't very concerned about this – when you set the goal line, you see pretty easily that it's attached to the `Count` column. If user wants to use a different column for the goal, the user can simply swap the order of Y-axis columns.

**Backend doesn't know anything about stacking logic**

screen shot 2018-01-17 at 18 45 39

In this case the frontend stacks the y-axis column values on top of each other. The user probably expects that the goal is compared against the sum of column values. This doesn't currently happen in the backend. We thought that this is a pretty rare scenario and also not worth the effort – although I don't know should we disable alerts for this scenario in the frontend.

If you choose `Stack - 100%` option then having a goal line doesn't make any sense because the top value of y axis is always `100%` regardless of column values.

--- 
@senior will also do some testing related to scenarios when the Y axis column/columns are set manually (via `graph.metrics` array in visualization settings object) because he wasn't confident whether those scenarios are treated correctly at the moment.

**Comment by attekei** [comment]
Now alerts are disabled in frontend if the question has >= explicitly set Y-axis columns as we discussed with @mazameli.

@senior I think that we should add this case to the logic that determines whether the alerts of question should be removed after saving a question. In JS the condition this case would be something like `card.visualization_settings["graph.metrics"] && card.visualization_settings["graph.metrics"].length > 1`

**Comment by senior** [comment]
@attekei It's a good thing we talked yesterday. Looking for "graph.metrics" is broken now. I have recreated the issue and fixed it locally. I'll push up a commit that fixes it once I've got a test put together.

I can add that additional alert clearing condition. Just to make sure I'm clear. If we had an alert with one graph.metric, then it gets updated to have two (or more), I should delete the alert?

**Comment by attekei** [comment]
>  If we had an alert with one graph.metric, then it gets updated to have two (or more), I should delete the alert?
There might be either one `graph.metric` or `graph.metric` might not be present at all. And yes, if there ever is two or more metrics, then the alert should be removed if there is one :)

**Comment by senior** [comment]
@attekei added a commit for pulses to use the metrics and dimensions column specified

File: frontend/src/metabase-lib/lib/Question.js [hunk 1]
```diff
@@ -231,17 +231,28 @@ export default class Question {
...
+        const isLineAreaBar = display === "line" || display === "area" || display === "bar"
+
         if (display === "progress") {
             return ALERT_TYPE_PROGRESS_BAR_GOAL;
-        } else if (mode && mode.name() === "timeseries") {
+        } else if (isLineAreaBar) {
             const vizSettings = this.card().visualization_settings;
+
+            // Don't support multiseries questions
+            // FIXME Atte Keinänen 1/18/17: This only detects visualizations where the user has manually
```
Comment (senior) [inline frontend/src/metabase-lib/lib/Question.js]: Is this something that you're intending to fix in this PR? I don't really know what the context of the FIXME is and what is missing if it's not there.

**Review by senior — COMMENTED** [review]
I've added a commit to this PR that will recognize user-specified X and Y axis values from the visualization settings

**Review by mazameli — APPROVED** [review]
This all appears to be working correctly to me. Two notes: 1) I could only get this to work with questions saved after I pulled the latest changes; I couldn't get it to work with previously saved questions. 2) One thing I haven't verified is that we don't send alerts for multiseries questions; waiting until the next hourly trigger to make sure of that.

**Comment by mazameli** [comment]
I verified that we're indeed not sending goal alerts for multi-series time series questions.

However, there are some UX warts with this:
- Why do we let the user set up that alert to begin with if we're never going to send the alert?
- If we were to prevent goal-based alerts from being available for multi-series time series questions, would it be clear to the user why that's the case?

**Comment by attekei** [comment]
> Why do we let the user set up that alert to begin with if we're never going to send the alert?

We definitely shouldn't, and it's something I should still work on (will take a while to implement).

(@senior the FIXME comment was related to this; I did earlier the quickest possible thing and disabled alerts if there are multiple user-set Y-axis columns. I didn't disable alerts yet if there are no user-set Y-axis columns but the "default column logic" results in multiple y-axis columns).

> If we were to prevent goal-based alerts from being available for multi-series time series questions, would it be clear to the user why that's the case?

Not really. The whole alerts UX experience isn't currently very explicit about why a specific type of alert is available or why alerts aren't aren't available at all for a given question.

---

It's unclear to me whether we want to include this PR and https://github.com/metabase/metabase/pull/6749#event-1437539405 to 0.28 – I can prioritize this if needed. Ping @salsakran

File: frontend/src/metabase/dashboard/components/AddSeriesModal.jsx [hunk 2]
```diff
@@ -212,7 +212,7 @@ export default class AddSeriesModal extends Component {
                     <div className="flex-full ml2 mr1 relative">
                         <Visualization
                             className="spread"
-                            series={series}
+                            rawSeries={series}
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/components/AddSeriesModal.jsx]: What's the rationale for renaming this `rawSeries`?

File: frontend/src/metabase/query_builder/selectors.js [hunk 3]
```diff
@@ -2,6 +2,12 @@
 import { createSelector } from "reselect";
 import _ from "underscore";
 
+// Needed due to wrong dependency resolution order
+// eslint-disable-next-line no-unused-vars
+import Visualization from "metabase/visualizations/components/Visualization";
```
Comment (tlrobinson) [inline frontend/src/metabase/query_builder/selectors.js]: You can just do `import "metabase/...";`

**Review by tlrobinson — COMMENTED** [review]
Just a few things, plus what we talked about regarding moving the defaults computation to on save.

File: frontend/src/metabase-lib/lib/Alert.js [hunk 4]
```diff
@@ -7,8 +7,8 @@ export type AlertType =
     | ALERT_TYPE_TIMESERIES_GOAL
     | ALERT_TYPE_PROGRESS_BAR_GOAL;
 
-export const getDefaultAlert = (question, user) => {
-    const alertType = question.alertType();
+export const getDefaultAlert = (question, user, visualizationSettings) => {
+    const alertType = question.alertType(visualizationSettings);
```
[the remark this answers is not in the document]
Comment (attekei) [reply frontend/src/metabase-lib/lib/Alert.js]:
```
    /**
     * Returns the type of alert that current question supports
     *
     * The `visualization_settings` in card object doesn't contain default settings,
     * so you can provide the complete visualization settings object to `alertType`
     * for taking those into account
     */
    alertType(visualizationSettings) {
```

File: frontend/src/metabase/dashboard/components/AddSeriesModal.jsx (hunk 2 above)
Comment (attekei) [reply frontend/src/metabase/dashboard/components/AddSeriesModal.jsx]: See these new selectors:

https://github.com/metabase/metabase/blob/3e39686537be7c3ae2104078fbc67f52381a2d52/frontend/src/metabase/query_builder/selectors.js#L173-L200

Basically I wanted to have a distinction between an untransformed and a transformed series object. Maybe `rawSeries` isn't the best name for it, but I think it's good that now also in `Visualization` there isn't duplicate meaning for `series`.

File: frontend/src/metabase/query_builder/selectors.js (hunk 3 above)
Comment (attekei) [reply frontend/src/metabase/query_builder/selectors.js]: True!

**Comment by attekei** [comment]
I talked with @tlrobinson how to proceed and now I'm planning the following:

1) Default x/y axis columns will be included to the card object always when the card is saved. 

2) Frontend will always save the question prior to creating/saving an alert. This way we can be certain that even for older questions goal alerts will use correct y axis columns for comparing the values against the goal line.

3) This all requires extracting question creation/saving logic to Redux action, something that has been forgotten to do earlier.

**Comment by attekei** [comment]
I implemented what I planned above and fixed some older issues along the way.

After tests pass in CI I think this is mergeable. I haven't tested yet if the complete goal alert creation + triggering flow works (it's slow because you have to always wait until the current hour changes...)

**Comment by senior** [comment]
@attekei What I do rather than wait for the job to trigger again is to change the time the job runs and restart. You can do this by changing the cron string on [this line](https://github.com/metabase/metabase/blob/master/src/metabase/task/send_pulses.clj#L74) from "0 0 * * * ? *" to "0 <minutes> * * * ? *" where `<minutes>` is the time you want it to run i.e. 12. Then restart the backend and it will run at 22 minutes past the hour rather than right on the hour.

I've thought about added a debugging endpoint that just always runs the job when you GET it, but I haven't added that yet.

**Comment by attekei** [comment]
Ah true, forgot that I can directly modify the cron strings! Thanks for the tip!

**Comment by salsakran** [comment]
Would it make sense for the cron to be `* * * * * ? *` in dev-mode?

**Comment by mazameli** [comment]
@attekei it seems this is still missing the small UX fix we talked about implementing for multi-series time series. Otherwise this seems good to go from my perspective (besides the currently failing tests of course ;) )

**Comment by senior** [comment]
@salsakran For me, it's probably be too chatty to be useful. I typically have a few test pulses sitting around in my dev db and most of the time I'm not testing them. As it sits now I'll occasionally get a few emails and I just delete them, but getting them every second or every minute would probably getting annoying in my current dev flow.

**Pull request merged**